service/history/hsm/operation_logger.go

```diff
@@ -0,0 +1,349 @@
+	}
+	if config.Logger == nil {
+		return nil, fmt.Errorf("logger must be provided")
+	}
+
+	l := &OpLog{
+		operations:    make([]*persistencespb.StateMachineOperation, 0),
+		maxOperations: config.MaxOperations,
+		retention:     config.Retention,
+		pruneInterval: config.PruneInterval,
+		stopCh:        make(chan struct{}),
+		logger:        config.Logger,
+	}
+
+	ready := make(chan struct{})
+	errCh := make(chan error, 1) // Buffer of 1 to avoid goroutine leak if initialization fails
+
+	go func() {
+		defer close(errCh)
+		close(ready)
+		l.pruneLoop()
+	}()
+
+	<-ready
+
+	config.Logger.Debug("Initializing operation log",
+		tag.Name("oplog"),
+		tag.Value("initialize"))
+
+	return l, nil
+}
+
+func (l *OpLog) LogOperation(ctx context.Context, op *persistencespb.StateMachineOperation) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	l.Lock()
+	defer l.Unlock()
+
+	if l.closed {
+		return fmt.Errorf("operation log is closed")
+	}
+
+	l.logger.Debug("Logging operation",
+		tag.Name("oplog"),
+		tag.Value(op.OperationType.String()))
+
+	opCopy := proto.Clone(op).(*persistencespb.StateMachineOperation)
+
+	// If we're at capacity, remove oldest entries
+	if len(l.operations) >= l.maxOperations {
+		l.logger.Warn("Operation log at capacity, dropping oldest entry",
+			tag.Name("oplog"),
+			tag.Value("capacity_reached"))
+		l.operations = l.operations[1:]
+	}
+
+	l.operations = append(l.operations, opCopy)
+	return nil
+}
+
+// TODO: Consider memory allocation optimizations:
+// - Consider returning an iterator instead
+func (l *OpLog) GetOperations(ctx context.Context, path []string) ([]*persistencespb.StateMachineOperation, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	l.RLock()
+	defer l.RUnlock()
+
+	if l.closed {
+		return nil, fmt.Errorf("operation log is closed")
+	}
+
+	l.logger.Debug("Fetching operations",
+		tag.Name("oplog"),
+		tag.Value(fmt.Sprintf("path: %v", path)))
+
+	// Pre-allocate with reasonable size
+	matches := make([]*persistencespb.StateMachineOperation, 0, min(len(l.operations), maxResultsLimit))
+
+	// Break up work into chunks and check context periodically
+	const chunkSize = 1000
+	for i := 0; i < len(l.operations) && len(matches) < maxResultsLimit; i += chunkSize {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		end := i + chunkSize
+		if end > len(l.operations) {
+			end = len(l.operations)
+		}
+
+		for _, op := range l.operations[i:end] {
+			if pathEquals(op.Path, path) {
+				matches = append(matches, proto.Clone(op).(*persistencespb.StateMachineOperation))
+				if len(matches) >= maxResultsLimit {
+					l.logger.Warn("GetOperations reached result limit",
+						tag.Name("oplog"),
+						tag.Value(fmt.Sprintf("limit: %d", maxResultsLimit)))
+					break
+				}
+			}
+		}
+	}
+
+	return matches, nil
+}
+
+func (l *OpLog) PruneOperations(ctx context.Context, retention time.Duration) error {
```

Comment:
You can delete this, it's not a requirement for this task, we don't want or need to prune operations.

Reply:
Removed